Removed all formatting issues. Left in the spring converted file abstraction for now.

garpinc opened the pull request:
Not sure how to make the spring abstractions optional. The code as it stands requires it.. I'd need to do some refactoring and it wouldn't be clean since i'd be repeating code in places.. It does work the way it is now but if you have some ideas let me know.

leleuj commented:
pac4j is made for being multi-frameworks, it means it cannot have specific dependencies on Spring (or Play or Undertow...) So the main target should be to change pac4j to become open to extensions required for these dependencies.

You can try to go further by having the Spring dependency as optional. Example for Kryo we already have:
```
<dependency>
  <groupId>com.esotericsoftware</groupId>
  <artifactId>kryo-shaded</artifactId>
  <scope>compile</scope>
  <optional>true</optional>
</dependency>
```
That said, there is another caveat: classes using the Spring dependency should be new ones separated from the existing ones to prevent pac4j users from being tied to the Spring dependency. So for example, no Spring package in the imports of the `CommonHelper` class.

This is maybe not achievable, though.

garpinc commented:
So in that case what's the issue with copied classes?
________________________________
From: LELEU Jérôme<mailto:notifications@github.com>
Sent: ‎3/‎9/‎2016 4:30 AM
To: pac4j/pac4j<mailto:pac4j@noreply.github.com>
Cc: garpinc<mailto:garpinc@hotmail.com>
Subject: Re: [pac4j] Removed all formatting issues. Left in the spring converted file abstraction for now. (#519)

pac4j is made for being multi-frameworks, it means it cannot have specific dependencies on Spring (or Play or Undertow...) So the main target should be to change pac4j to become open to extensions required for these dependencies.

You can try to go further by having the Spring dependency as optional. Example for Kryo we already have:
```
<dependency>
  <groupId>com.esotericsoftware</groupId>
  <artifactId>kryo-shaded</artifactId>
  <scope>compile</scope>
  <optional>true</optional>
</dependency>
```
That said, there is another caveat: classes using the Spring dependency should be new ones separated from the existing ones to prevent pac4j users from being tied to the Spring dependency. So for example, no Spring package in the imports of the `CommonHelper` class.

This is maybe not achievable, though.



---
Reply to this email directly or view it on GitHub:
https://github.com/pac4j/pac4j/pull/519#issuecomment-194205548

leleuj commented:
Copying classes is not clean. Imagine someone else does the same for another framework. pac4j would become a patch work of copied classes.

What will happen when you effectively use the Spring dependency? Which classes will be used? What happens when you upgrade the dependency?

pac4j being an engine, the best you can have is extensible classes you can easily use or specific classes with an optional dependency.

garpinc commented:
The problem for the most part is that the pac4j code affected are interfaces.. Unless I make the objects being passed around just a java Object there is going to have to some pac4j interface that abstracts the file handling. If we can’t use the spring interfaces directly because you don’t want the dependency then I think the only real way is the way I implemented it.. A mirror of the interfaces (because the interfaces are good representations of the abstraction) in the pac4j world.. I could make the implementation FileSystemResource use the actual spring implementation in an optional way. But I think we need the 2 interfaces Resource and WritableResource in the pac4j world..

 

Would that suffice?

garpinc commented:
Hey.. So I think really the best way going forward is to have an abstraction on these files but as you say it doesn't have to be a copy of spring files.. So basically what I did was make bare minimal interfaces and one implementation that are pac4j abstractions.. I hate continuing to merge with master.. I hope you will accept my pull request

leleuj commented on pac4j-core/src/main/java/org/pac4j/core/client/Clients.java:
```diff
@@ -104,7 +105,10 @@ protected void internalInit() {
     public Client findClient(final WebContext context) {
         final String name = context.getRequestParameter(this.clientNameParameter);
-        CommonHelper.assertNotBlank("name", name);
+		if (name == null && defaultClient != null) {
```
What is this `defaultClient` for? It doesn't seem related to the resource changes...

garpinc replied:
basically if on callback url there is no clientName specified you can configure in a default client., It's not related to resource changes.. I'm just not sure how to make that change separate at this point.

leleuj replied:
For a different need, we need a different pull request. Recommit the master file version.
Notice this must be discussed before submitting a pull request: do we really want to bring in that behaviour? I'm really not sure.

leleuj commented on pac4j-core/src/main/java/org/pac4j/core/io/FileSystemResource.java:
```diff
@@ -0,0 +1,40 @@
+import java.io.InputStream;
+import java.io.OutputStream;
+
+public class FileSystemResource implements WritableResource {
```
Some Javadoc will be appreciated.

garpinc replied:
Added some javadoc

leleuj replied:
It might be the case for other classes.

leleuj commented on pac4j-core/src/main/java/org/pac4j/core/io/FileSystemResource.java:
```diff
@@ -0,0 +1,40 @@
+
+	@Override
+	public boolean exists() {
+		// TODO Auto-generated method stub
```
I'd prefer not to keep any `TODO`. My feeling is that either you throw some exception to say it is not implemented or you implement the real logic.

garpinc replied:
Agreed.. implemented for next checkin

leleuj commented on pac4j-core/src/main/java/org/pac4j/core/io/InputStreamSource.java:
```diff
@@ -0,0 +1,24 @@
+ *
+ * @see java.io.InputStream
+ * @see Resource
+ */
```
I like to have the `@since` annotation for history.

garpinc replied:
Added it for next checkin

leleuj replied:
After the hard work you are doing, I guess you don't want to forget the `author` annotation as well. I think I saw the problem in several other files.

leleuj commented on pac4j-core/src/main/java/org/pac4j/core/io/WritableResource.java:
```diff
@@ -0,0 +1,40 @@
+/*
```
Please remove the header.

leleuj commented on pac4j-http/src/main/java/org/pac4j/http/client/indirect/FormClient.java:
```diff
@@ -79,23 +79,36 @@ protected UsernamePasswordCredentials retrieveCredentials(final WebContext conte
-                final String message = "Username and password cannot be blank -> return to the form with error";
-                logger.debug(message);
-                throw RequiresHttpAction.redirect(message, context, redirectionUrl);
+				// it's an AJAX request -> unauthorized (instead of a
```
What is this change? You shouldn't mix different changes in the same pull request.

garpinc replied:
This change addresses a bug I saw when a login to the authentication URL was accessed via an ajax call. When there is an ajax call you can't do a redirect because the browser will redirect before the ajax caller get's control. Hence this fixes that and for ajax it doesn't do a redirect.

leleuj replied:
OK. If there is a problem, we'll fix it, but in another pull request. Just recommit the original current master file and the change will go away from the pull request. We'll deal with that problem later on. You can already open a Github issue so we don't forget about it.

leleuj commented on pac4j-saml/src/main/java/org/pac4j/saml/client/SAML2Client.java:
```diff
@@ -150,24 +150,24 @@ protected void initSAMLContextProvider(final MetadataResolver metadataManager) {
+		String finalCallbackUrl = computeFinalCallbackUrl(context);
+		this.spMetadataResolver = new SAML2ServiceProviderMetadataResolver(this.configuration.getServiceProviderMetadataPath(),
+				this.configuration.getServiceProviderMetadataResource(), finalCallbackUrl,
+				(this.configuration.getServiceProviderEntityId().equals("CALLBACK_URL") ? finalCallbackUrl
```
What is this test for?

garpinc replied:
In your https://github.com/pac4j/spring-security-pac4j-demo/blob/master/src/main/resources/securityContext.xml you had 

> <property name="serviceProviderEntityId" value="http://localhost:8080/callback?client_name=SAML2Client" />

I basically thought instead I could just refer to CALLBACK_URL and in those cases it would just put the callback url rather than hardcoding the config defined somewhere else.

leleuj replied:
I get the idea, but can you remove this for now? Let's try to keep one change at a time, there are already many things to account for.

leleuj commented on pac4j-saml/src/main/java/org/pac4j/saml/client/SAML2ClientConfiguration.java:
```diff
@@ -59,17 +69,45 @@ public SAML2ClientConfiguration() {}
+    }
+
+	private SAML2ClientConfiguration(final KeyStore keyStore, final String keyStoreAlias, final String keyStoreType,
+			                         final Resource keystoreResource, final String keystorePath, final String keystorePassword,
```
Why do we have a `keystoreResource` and a `keystorePath`?

garpinc replied:
Mainly I wanted all your original code paths to still work so you could pass either the original keystorePath or the new keystoreResource. I figured someone in the wild may have made a call to the old methods and I didn't want my changes to break it. Is this not desirable?

leleuj replied:
It depends. Here, I don't understand why we have **at the same time** both parameters: isn't it one or the other?

garpinc replied:
It's a private method used by both external public methods that only allow application of one or other.. Then in the private method I make sure that one or the other is supplied.

leleuj commented on pac4j-saml/src/main/java/org/pac4j/saml/sso/impl/SAML2DefaultResponseValidator.java:
```diff
@@ -674,7 +674,12 @@ protected final void validateSignature(final Signature signature, final String i
         final DateTime after =  DateTime.now().minusSeconds(acceptedSkew + interval);
-        return issueInstant.isBefore(before) && issueInstant.isAfter(after);
+        boolean isDateValid = issueInstant.isBefore(before) && issueInstant.isAfter(after);
+		if (logger.isTraceEnabled() && !isDateValid) {
```
We don't need a `logger.isTraceEnabled` here.

garpinc replied:
I always do that for performance reasons: http://learnertobeginner.blogspot.com/2010/06/how-to-use-isdebugenabled-effectly.html. Shall I remove it as requested?

leleuj replied:
Yes, you should remove it since slf4j version 1.7 I think, but you also need to change the way you log things: from `logger.trace("myvalue is: " + myvalue)` to `logger.trace("myvalue: {}", myvalue);`. Following this rule, there will be no performance impact (no String building, these are just parameters).

leleuj commented on pac4j-core/src/main/java/org/pac4j/core/util/CommonHelper.java:
```diff
@@ -226,20 +242,83 @@ public static String toString(final Class<?> clazz, final Object... args) {
+			prefix = name.substring(0, prefixEnd);
+			path = name.substring(prefixEnd + 1);
+		}
+		if (prefix == null) {
```
It's not backward compatible, we support the case where we have no prefix.

garpinc replied:
I added the if null for findbugs which told me that the prefix maybe null.. I'll take it out and see if travis will build

leleuj commented on pac4j-core/src/main/java/org/pac4j/core/util/CommonHelper.java:
```diff
@@ -226,20 +242,83 @@ public static String toString(final Class<?> clazz, final Object... args) {
+			return getInputStreamViaHttp(name);
+		case HTTPS_PREFIX:
+			return getInputStreamViaHttp(name);
+		default:
```
And you have a `default` case...

garpinc replied:
I added the if null for findbugs which told me that the prefix maybe null.. I'll take it out and see if travis will build

leleuj replied:
Findbugs may be "wrong" or too picky sometimes and we can exclude applied rules if it makes sense.
Here, we must be backward compatible for all users using this without a prefix and this is certainly the majority of the pac4j users.

I'm surprised that no test fails on this: we lack tests on this (change).

leleuj commented on pac4j-core/src/main/java/org/pac4j/core/util/CommonHelper.java:
```diff
@@ -226,20 +242,83 @@ public static String toString(final Class<?> clazz, final Object... args) {
+
+	private static InputStream getInputStreamViaHttp(String name) {
+		URLConnection con = null;
+		try {
```
Very strange: no `finally` to close the connection?

garpinc replied:
Caller is responsible for closing inputstream. I added that to comments.

leleuj replied:
OK. How the connection is closed if no exception occurs?

garpinc replied:
Once again the job of this is not to close the stream.. the caller has to enclose the call to this method with the finally to close the stream.

leleuj replied:
I was mistaken, I meant: how the connection is supposed to be disconnected in regular use case (no error)? By the caller?

garpinc replied:
From http://docs.spring.io/autorepo/docs/spring/3.2.x/spring-framework-reference/html/resources.html  "It is expected that each invocation returns a fresh InputStream. It is the responsibility of the caller to close the stream.". Here are some examples in regular use case closing the stream: http://www.mkyong.com/spring/spring-resource-loader-with-getresource-example/

leleuj replied:
I agree: `inputStream.close();` must be handled by the caller. But how the `con.disconnect()` is called?

garpinc replied:
http://kingori.co/minutae/2013/04/httpurlconnection-disconnect/ and http://stackoverflow.com/questions/4767553/safe-use-of-httpurlconnection.. Basically this code is the same as it is in spring.. When the inputstream is close it can free up resources and depending on keep-alive setting will do the disconnect as far as I understand it.

leleuj replied:
OK

leleuj commented on pac4j-core/src/test/java/org/pac4j/core/util/CommonHelperTests.java:
```diff
@@ -146,6 +146,11 @@ public void testAreEqualsIgnoreCaseAndTrimFails() {
     }
+    
+    @Test
+    public void testClassPathPrefix() {
```
Just one test for all these changes.

garpinc replied:
Mainly I thought your other tests were covering most of my code changes already especially since the paths are now converted to resources in code.

leleuj replied:
It should, but nothing is perfect. I think there is a lack of tests here: we should have ones for all prefixes and for the use case with no prefix.

leleuj commented on pac4j-core/src/main/java/org/pac4j/core/io/Resource.java:
```diff
@@ -0,0 +1,32 @@
+ * certain resources. The actual behavior is implementation-specific.
+ *
+ */
+public interface Resource extends InputStreamSource {
```
I don't think we use `InputStreamSource` solely: can we merge it with `Resource`?

garpinc replied:
WriteableResource extends Resource so I don't think they should be merged for that reason. Resource is the common interface, input is for read and writeable is for write

leleuj replied:
But no other interface extends from `InputStreamSource` directly, doesn't it? So in our case, a `Resource` is necessarily a `InputStreamSource` isn't it? We never distinguish between both interfaces.
Or do you directly use the `InputStreamResource` for your Spring extension?
That said, it might be cleaner to keep both for clarity.

leleuj commented on pac4j-core/src/main/java/org/pac4j/core/util/CommonHelper.java:
```diff
@@ -226,20 +242,83 @@ public static String toString(final Class<?> clazz, final Object... args) {
+		case CLASSPATH_PREFIX:
+			return Thread.currentThread().getContextClassLoader().getResourceAsStream(path);
+		case HTTP_PREFIX:
+			logger.warn("IdP metadata is retrieved from an insecure http endpoint [{}]", path);
```
This SAML log has nothing to do here, in this generic `CommonHelper` method.

leleuj commented on pac4j-core/src/main/java/org/pac4j/core/util/CommonHelper.java:
```diff
@@ -226,20 +242,83 @@ public static String toString(final Class<?> clazz, final Object... args) {
+		}
+	}
+
+	public static Resource getResource(final String idpMetadataPath) {
```
`idMetadataPath`:  this seems specific to SAML.

leleuj commented on pac4j-saml/src/main/java/org/pac4j/saml/client/SAML2ClientConfiguration.java:
```diff
@@ -19,6 +22,9 @@
+
+	private Resource keystoreResource;
 
     private String keystorePath;
```
Do we still really need a `keystorePath`? Cannot it be some encapsulated data of a file resource for the `keystoreResource`?

garpinc replied:
Mainly I wanted all your original code paths to still work so you could pass either the original keystorePath or the new keystoreResource. I figured someone in the wild may have made a call to the old methods and I didn't want my changes to break it. Is this not desirable?

leleuj replied:
Sure. But, I imagine you can remove the internal `keystorePath ` property and keep the setter (and getter) to initialize a specific file resource in the `keystoreResource` based on the provided path: it would be better: backward compatible and cleaner a the same time.

leleuj commented:
I started reviewing the changes and stopped as there are already many comments. It's getting better though.

First of all, this pull request should only target the introduction of the `Resource` concept. Nothing more. No changes in `Clients` or `FormClient`. I see many side-effects changes as well. These must be justified or removed.

Regarding the core changes, I'm in line with adding new interfaces. It seems that `InputStreamResource` could be merged in `Resource` as it's not used anywhere, but maybe you need it for Spring extensions. See my different comments...

garpinc commented:
Also regarding InputStreamResource. I do need it for spring extensions

leleuj commented:
OK. So let's keep `InputStreamResource`. I replied to all your comments, I see that some of my remarks have not been accounted for so far...

garpinc commented:
I refactored my end and removed InputStreamResource per your request, removed various checkins to re-add in a future pull request, and added the test cases for the prefixes. let me know how this iteration is

leleuj commented:
Thanks for your tenacity on this. I will review it shortly.

leleuj commented on pac4j-core/src/main/java/org/pac4j/core/util/CommonHelper.java:
```diff
@@ -19,7 +26,16 @@
+
+	public static final String RESOURCE_PREFIX = "resource";
+	public static final String CLASSPATH_PREFIX = "classpath";
+	public static final String HTTP_PREFIX = "http";
```
In fact, we already these constants in the `HttpConstants` interface.

leleuj commented on pac4j-saml/src/main/java/org/pac4j/saml/client/SAML2ClientConfiguration.java:
```diff
@@ -59,25 +67,56 @@ public SAML2ClientConfiguration() {}
+        }
         this.keystorePassword = keystorePassword;
         this.privateKeyPassword = privateKeyPassword;
+		this.identityProviderMetadataResource = identityProviderMetadataResource;
```
Cannot we do like for the `keystorePath` and `keystoreResource` regarding the `identityProviderMetadataPath` and `identityProviderMetadataResource`?

leleuj commented on pac4j-saml/src/main/java/org/pac4j/saml/crypto/KeyStoreCredentialProvider.java:
```diff
@@ -43,14 +48,43 @@
+			final InputStream inputStream;
+			if (keyStoreResource != null) {
+				try {
+					inputStream = keyStoreResource.getInputStream();
```
I don't see where this stream is closed...

garpinc replied:
It's closed in the loadKeyStore method.. I didn't change that.

leleuj commented on pac4j-saml/src/main/java/org/pac4j/saml/crypto/KeyStoreCredentialProvider.java:
```diff
@@ -43,14 +48,43 @@
+			if (keyStoreResource != null) {
+				try {
+					inputStream = keyStoreResource.getInputStream();
+				} catch (IOException e) {
```
No `finally` to close the stream?

garpinc replied:
I agree there should be a finally I'll add it..

leleuj commented on pac4j-saml/src/main/java/org/pac4j/saml/crypto/KeyStoreCredentialProvider.java:
```diff
@@ -43,14 +48,43 @@
+					throw new TechnicalException(e);
+				}
+			} else {
+				inputStream = CommonHelper.getInputStreamFromName(keyStorePath);
```
No `try/catch IOException` here?

leleuj commented on pac4j-saml/src/main/java/org/pac4j/saml/crypto/KeyStoreCredentialProvider.java:
```diff
@@ -114,12 +148,17 @@ private KeyStore loadKeyStore(final InputStream inputStream, final String storeP
-            }
+			while (aliases.hasMoreElements()) {
+	            String currentAlias = aliases.nextElement();
+				if (keyStoreAlias == null) {
```
I'm not sure to understand this use case.

garpinc replied:
Previously if the keystore had more than one alias you just picked the first one. I have a scenario where my keystore contains more than one alias and the one i need for saml isn't always the first one so if you specify which alias then it uses the correct one.

leleuj replied:
OK

leleuj commented on pac4j-saml/src/main/java/org/pac4j/saml/crypto/KeyStoreCredentialProvider.java:
```diff
@@ -114,12 +148,17 @@ private KeyStore loadKeyStore(final InputStream inputStream, final String storeP
+	            String currentAlias = aliases.nextElement();
+				if (keyStoreAlias == null) {
+					return currentAlias;
+				} else if (currentAlias.equals(keyStoreAlias)) {
```
I think this condition could be merge with the previous one.

garpinc replied:
Not sure I understand what you mean here.

leleuj replied:
In both cases, you return the `currentAlias`, you could have one condition with `||`.

leleuj commented:
+1 Looks good to me with a few comments to address before the merge

Waiting also for @mmoayyed 's feedback on the SAML changes...

garpinc commented:
I have addressed the issues you mentioned.. Also re-implemented these changes on 1.8.x pull request. Will keep those in sync with your comments here.

leleuj commented on pac4j-core/src/main/java/org/pac4j/core/io/Resource.java:
```diff
@@ -0,0 +1,48 @@
+ * physical form, but a URL or File handle can just be returned for
+ * certain resources. The actual behavior is implementation-specific.
+ * 
+ *  @since 1.9.0
```
I'd like to see the `@author` annotation here also.

leleuj commented:
A few comments more, but it looks good globally.

@mmoayyed : is it ok for you?

leleuj commented on pac4j-core/pom.xml:
```diff
@@ -45,6 +45,12 @@
             <groupId>org.springframework</groupId>
+            <artifactId>spring-core</artifactId>
+            <scope>compile</scope>
+            <version>${spring.version}</version>
```
It should be marked with `<optional>true</optional>`  because I don't want to pull the Spring dependency for everybody.

leleuj commented on pac4j-core/src/main/java/org/pac4j/core/io/SpringResourceWrapper.java:
```diff
@@ -0,0 +1,34 @@
+ * A wrapper that adapts a spring {@link org.springframework.core.io.Resource} to a pac4j {@link Resource}.
+ * 
+ * @author Keith Garry Boyce
+ * @since 1.4.3
```
Version 1.9.0.

leleuj commented on pac4j-core/src/main/java/org/pac4j/core/io/SpringWritableResourceWrapper.java:
```diff
@@ -0,0 +1,25 @@
+ * A wrapper that adapts a spring {@link org.springframework.core.io.WritableResource} to a pac4j {@link WritableResource}.
+ * 
+ * @author Keith Garry Boyce
+ * @since 1.4.3
```
Version 1.9.0

leleuj commented:
Let's fix my latest comments before I merge.

leleuj commented on pac4j-saml/src/main/java/org/pac4j/saml/client/SAML2Client.java:
```diff
@@ -150,7 +150,7 @@ protected void initSAMLContextProvider(final MetadataResolver metadataManager) {
 
     protected MetadataResolver initServiceProviderMetadataResolver(final WebContext context) {
-        this.spMetadataResolver = new SAML2ServiceProviderMetadataResolver(this.configuration.getServiceProviderMetadataPath(),
+		this.spMetadataResolver = new SAML2ServiceProviderMetadataResolver(this.configuration.getServiceProviderMetadataPath(),
```
I think that passing the `configuration` could be sufficient, couldn't it? To mirror what's done for the identity provider...

leleuj commented on pac4j-saml/src/main/java/org/pac4j/saml/metadata/SAML2IdentityProviderMetadataResolver.java:
```diff
@@ -41,16 +38,33 @@
     protected final static String HTTP_PREFIX = "http";
```
These two constants are no longer necessary.

leleuj commented:
Thank your very much for your work. Thanks for your patience and tenacity through this long merging process.
